/*
    Un Product è un prodotto registrato dal Business Admin come producibile all'interno della Company. 
    Si conosce il nome del prodotto (univoco), la company a cui appartiene, chi lo ha registrato, informazioni significative sul prodotto,
    la tipologia (se raccolto, elaborato, ecc.).
    Nel caso in cui sia prodotto in un'azienda di produzione e sia di Harvested allora si consoce anche il campo di produzione
*/

syntax = "proto3";

message Product {
    
    // Documento, contiene un riferimento (url) allo storage off-chain e l'hash per l'integrità
    message Document {
        string url = 1;

        string hash = 2;

        string description = 3;
    }

    // Informazioni utili
    message Info {
        string description = 1;

        repeated Document documents = 2;
    }

    // Tipologia del prodotto
    enum ProductType {
        HARVESTED = 0; //es. olive
        PROCESSED = 1; //es. olio
        FINAL = 2; //es. bottiglia d'olio
    }
    
    // Nome prodotto (univoco)
    string name = 1;

    // Azienda produttrice
    string company = 2;

    // Informazioni utili
    Info info = 3;

    // Tipologia del prodotto (condiziona batch ed eventi)
    ProductType type = 4;

    // Id del campo di produzione (se type = harvested)
    string field = 5;

    // Timestamp
    uint64 timestamp = 6;
}

message ProductContainer {
    repeated Product entries = 1;
}